libraries/abstractions/pkcs11/utest/iot_pkcs11_mbedtls_utest.c

```diff
@@ -171,112 +299,540 @@ void test_pkcs11_C_InitializeInitTwice( void )
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
 
+    xResult = C_OpenSession( 0, xFlags, NULL, 0, NULL );
     TEST_ASSERT_EQUAL( CKR_ARGUMENTS_BAD, xResult );
+
+    xResult = C_OpenSession( 0, CKF_RW_SESSION, NULL, 0, &xSession );
+    TEST_ASSERT_EQUAL( CKR_SESSION_PARALLEL_NOT_SUPPORTED, xResult );
+
+    xResult = prvUninitializePkcs11();
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
 }
 
-/* ======================  END TESTING ============================ */
+/* ======================  TESTING C_OpenSession  ============================ */
+
+/*!
+ * @brief C_CloseSession happy path.
+ *
+ */
+void test_pkcs11_C_CloseSession( void )
+{
+    CK_RV xResult = CKR_OK;
+    CK_SESSION_HANDLE xSession = 0;
+
+    xResult = prvInitializePkcs11( ( SemaphoreHandle_t ) &xResult );
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
+
+    prvOpenSession( &xSession );
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
+
+    prvCloseSession( &xSession );
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
+
+    xResult = prvUninitializePkcs11();
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
+}
+/* ======================  TESTING C_Login  ============================ */
+
+/*!
+ * @brief C_Login happy path.
+ *
+ * @note This test will need to be updated if support is added for C_Login
+ *
+ */
+void test_pkcs11_C_Login( void )
+{
+    CK_RV xResult = CKR_OK;
+
+    xResult = C_Login( 0, 0, NULL, 0 );
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
+}
+
+/* ======================  TESTING C_CreateObject  ============================ */
+
+/*!
+ * @brief C_CreateObject Creating an EC private key happy path.
+ *
+ */
+void test_pkcs11_C_CreateObjectECPrivKey( void )
+{
+    CK_RV xResult = CKR_OK;
+    CK_SESSION_HANDLE xSession = 0;
+    CK_KEY_TYPE xPrivateKeyType = CKK_EC;
+    CK_OBJECT_CLASS xPrivateKeyClass = CKO_PRIVATE_KEY;
+    CK_BBOOL xTrue = CK_TRUE;
+    char * pucLabel = pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS;
+    /* DER-encoding of an ANSI X9.62 Parameters value */
+    CK_BYTE * pxEcParams = ( CK_BYTE * ) ( "\x06\x08" MBEDTLS_OID_EC_GRP_SECP256R1 );
+    CK_OBJECT_HANDLE xObject = 0;
+
+    /* Private value D. */
+    CK_BYTE * pxD = pvPkcs11MallocCb( EC_D_LENGTH, 1 );
+
+    TEST_ASSERT_NOT_EQUAL( NULL, pxD );
+
+    CK_ATTRIBUTE xPrivateKeyTemplate[] = EC_PRIV_KEY_INITIALIZER;
+
+    xResult = prvInitializePkcs11( ( SemaphoreHandle_t ) &xResult );
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
+
+    prvOpenSession( &xSession );
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
+
+    mbedtls_pk_init_CMockIgnore();
+    pvPortMalloc_Stub( pvPkcs11MallocCb );
+    PKCS11_PAL_FindObject_IgnoreAndReturn( 1 );
+    PKCS11_PAL_GetObjectValue_IgnoreAndReturn( CKR_OK );
+    mbedtls_pk_parse_key_IgnoreAndReturn( 0 );
+    PKCS11_PAL_GetObjectValueCleanup_CMockIgnore();
+    pvPortMalloc_Stub( pvPkcs11MallocCb );
+    mbedtls_ecp_keypair_init_CMockIgnore();
+    mbedtls_ecp_group_init_CMockIgnore();
+    mbedtls_ecp_group_load_IgnoreAndReturn( 0 );
+
+    /* current callstack */
```

Comment:
What does this comment mean?

Reply:
Sorry this was a tracker for me, forgot to remove it.